torch/lib/c10d/TcpStore.cpp

```diff
@@ -0,0 +1,326 @@
+      it++;
+    } else {
+      it = keys.erase(it);
+    }
+  }
+  return ret;
+}
+
+// TCPStore class methods
+TCPStore::TCPStore(const std::string& masterAddr,
+                   PortType masterPort,
+                   bool isServer)
+ : isServer_(isServer)
+ , tcpStoreAddr_(masterAddr)
+ , tcpStorePort_(masterPort)
+
+{
+  if (isServer_) {
+    // Opening up the listening socket
+    std::tie(masterListenSocket_, std::ignore) = tcputil::listen(masterPort);
+    // Now start the daemon
+    tcpStoreDaemon_ = std::unique_ptr<TCPStoreDaemon>(
+        new TCPStoreDaemon(masterListenSocket_)
+    );
+  }
+  // Connect to the daemon
+  storeSocket_ = tcputil::connect(tcpStoreAddr_, tcpStorePort_);
+}
+
+TCPStore::~TCPStore() {
+  ::close(storeSocket_);
+  if (isServer_) {
+    /**
+     * Store daemon should end because of closed connection.
+     * daemon destructor should join the thread
+     */
+    tcpStoreDaemon_.reset(nullptr);
+    ::close(masterListenSocket_);
+  }
+}
+
+void TCPStore::set(const std::string& key, const std::vector<uint8_t>& data) {
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::SET);
+  tcputil::sendString(storeSocket_, key, true);
+  tcputil::sendVector<uint8_t>(storeSocket_, data);
+}
+
+std::vector<uint8_t> TCPStore::get(const std::string& key) {
+  wait({key});
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::GET);
+  tcputil::sendString(storeSocket_, key);
+  return tcputil::recvVector<uint8_t>(storeSocket_);
+}
+
+int64_t TCPStore::add(const std::string& key, int64_t value) {
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::ADD);
+  tcputil::sendString(storeSocket_, key, true);
+  tcputil::sendValue<int64_t>(storeSocket_, value);
+  return tcputil::recvValue<int64_t>(storeSocket_);
+}
+
+bool TCPStore::check(const std::vector<std::string>& keys) {
+
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::CHECK);
+  SizeType nkeys = keys.size();
+  tcputil::sendBytes<SizeType>(storeSocket_, &nkeys, 1, (nkeys > 0));
+  for (size_t i = 0; i < nkeys; i++) {
+    tcputil::sendString(storeSocket_, keys[i], (i != (nkeys - 1)));
+  }
+  auto checkResponse = tcputil::recvValue<CheckResponseType>(storeSocket_);
+  if (checkResponse == CheckResponseType::READY) {
+    return true;
+  } else if (checkResponse == CheckResponseType::NOT_READY) {
+    return false;
+  } else {
+    throw std::runtime_error("ready or not_ready response expected");
+  }
+}
+
+void TCPStore::wait(
+    const std::vector<std::string>& keys,
+    const std::chrono::milliseconds& timeout) {
+
+  const auto start = std::chrono::steady_clock::now();
+  while (!check(keys)) {
+    const auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(
+        std::chrono::steady_clock::now() - start);
+    if (timeout != kNoTimeout && elapsed > timeout) {
+      throw std::runtime_error("Wait timeout");
+    }
+    /* sleep override */
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  }
```

Comment:
Perhaps this should just become the default implementation of `wait` for different stores? I think we're using the same strategy both in here and in the file one. Wdyt @pietern?

Reply:
We can do this as a stop gap but I would rather make this wake up to activity. For the TCPStore this means issuing a wait requests that returns as soon as it is fulfilled (the waiting part is done on the server side).

Reply:
That’s what the impl used to in THD (and before the last commit), so we might want to revert this part.

Reply:
The commit before last risked running out of sync between server and client. The notify/wakeup part can be reverted, but should hold on to the socket that expects a message for wakeup, not the keep_waiting/stop_waiting part.

Reply:
Ok, my bad. That must have been changed from THD, [which only has the `STOP_WAITING` command](https://github.com/pytorch/pytorch/blob/master/torch/lib/THD/base/data_channels/Store.cpp#L16).

Reply:
Ok, I changed it back.